### PR TITLE
MGMT-7601: V2 cluster update API

### PR DESCRIPTION
**Pull request by nmagnezi**
# Assisted Pull Request

## Description

Add v2 API for Cluster Update

## List all the issues related to this PR

- [ ] New Feature
- [ ] Bug fix
- [ ] Tests
- [ ] Documentation
- [ ] CI/CD 

## What environments does this code impact?

- [ ] Automation (CI, tools, etc)
- [ ] Cloud
- [ ] Operator Managed Deployments
- [x] None

## How was this code tested?

- [ ] assisted-test-infra environment
- [ ] dev-scripts environment
- [ ] Reviewer's test appreciated
- [ ] Waiting for CI to do a full test run
- [ ] Manual (Elaborate on how it was tested)
- [x] No tests needed

## Assignees

/cc @
/cc @

## Checklist

- [ ] Title and description added to both, commit and PR.
- [ ] Relevant issues have been associated (see [CONTRIBUTING] guide)
- [ ] Reviewers have been listed
- [ ] This change does not require a documentation update (docstring, `docs`, README, etc)
- [ ] Does this change include unit-tests (note that code changes require unit-tests)

## Reviewers Checklist

- [ ] Are the title and description (in both PR and commit) meaningful and clear?
- [ ] Is there a bug required (and linked) for this change?
- [ ] Should this PR be backported?

[Kubernetes community documentation]: https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#commit-message-guidelines
[CONTRIBUTING]: https://github.com/openshift/assisted-service/blob/master/CONTRIBUTING.md

### CODE REVIEW

**Comment by nmagnezi** [comment]
/hold
WIP

File: internal/bminventory/inventory.go [hunk 1]
```diff
@@ -2175,6 +2254,113 @@ func (b *bareMetalInventory) updateClusterInternal(ctx context.Context, params i
...
+	}
+
+	if err = b.noneHaModeV2ClusterUpdateValidations(cluster, params); err != nil {
+		log.WithError(err).Warnf("Unsupported update params in none ha mode")
+		return nil, common.NewApiError(http.StatusBadRequest, err)
+	}
+
+	if err = b.v2ValidateDNSDomain(*cluster, params, log); err != nil {
+		return nil, err
+	}
```
Comment (yevgeny-shnaidman) [inline internal/bminventory/inventory.go:2310]: what about DiskEncryption validation? Has it been moved to another location?

File: internal/bminventory/inventory.go [hunk 2]
```diff
@@ -2353,18 +2645,170 @@ func (b *bareMetalInventory) updateClusterData(_ context.Context, cluster *commo
...
+		dbReply := db.Model(&common.Cluster{}).Where("id = ?", cluster.ID.String()).Updates(updates)
+		if dbReply.Error != nil {
+			return common.NewApiError(http.StatusInternalServerError, errors.Wrapf(err, "failed to update cluster: %s", params.ClusterID))
+		}
+	}
+
+	return nil
+}
+
+func (b *bareMetalInventory) v2UpdateClusterData(_ context.Context, cluster *common.Cluster, params installer.V2UpdateClusterParams, usages map[string]models.Usage, db *gorm.DB, log logrus.FieldLogger, interactivity Interactivity) error {
```
Comment (yevgeny-shnaidman) [inline internal/bminventory/inventory.go:2659]: the same comment regarding DiskEncryption as above

File: internal/bminventory/inventory.go (hunk 1 above)
Comment (nmagnezi) [reply internal/bminventory/inventory.go]: Fixed

File: internal/bminventory/inventory.go (hunk 2 above)
Comment (nmagnezi) [reply internal/bminventory/inventory.go]: Fixed

**Comment by nmagnezi** [comment]
@yevgeny-shnaidman I also re-organized the commits so you can filter out the generated code by selecting the specific commit to view

File: internal/bminventory/inventory.go [hunk 3]
```diff
@@ -1981,46 +1981,58 @@ func (b *bareMetalInventory) noneHaModeClusterUpdateValidations(cluster *common.
 	return nil
 }
 
-func (b *bareMetalInventory) validateAndUpdateClusterParams(ctx context.Context, params *installer.UpdateClusterParams) (installer.UpdateClusterParams, error) {
+func (b *bareMetalInventory) v2NoneHaModeClusterUpdateValidations(cluster *common.Cluster, params installer.V2UpdateClusterParams) error {
```
Comment (yevgeny-shnaidman) [inline internal/bminventory/inventory.go:1984]: why do we need to rename this function? Why not leave it as is and just change the input params to be V2?

**Comment by rollandf** [comment]
/hold
Just saw @yevgeny-shnaidman comment

File: internal/bminventory/inventory.go (hunk 3 above)
Comment (nmagnezi) [reply internal/bminventory/inventory.go:1984]: This is the single function for which I have made a copy.
noneHaModeClusterUpdateValidations  updates `HostsRoles` which you do not have in `V2UpdateClusterParams`:
https://github.com/openshift/assisted-service/blob/bb4e8cf7fa547570ca84792d4eae72fa6a9f3f46/internal/bminventory/inventory.go#L1968-L1982

Therefore I needed a copy function that does everything else the same way (e.g. return codes for the other checks) but doesn't have that check.

Comment (nmagnezi) [reply internal/bminventory/inventory.go:1984]: and btw the `validateAndUpdateClusterParams` was not renamed. the GitHub diff is misleading here.
see the `validateAndUpdateClusterParams` function below at line 1996. the change there was `UpdateClusterParams` --> `V2UpdateClusterParams`

**Pull request merged**